### PR TITLE
[dashboards] Implement support for Geomap Dashboard Widget

### PR DESCRIPTION
**Pull request by DrkSephy**
This pull request implements support for creating Geomap widgets in Dashboards.

### CODE REVIEW

File: datadog/tests/resource_datadog_dashboard_geomap_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+package test
+
+import (
+	"testing"
+)
+
+const datadogDashboardGeomapConfig = `
+resource "datadog_dashboard" "geomap_dashboard" {
```
Comment (jbenais) [inline datadog/tests/resource_datadog_dashboard_geomap_test.go:8]: let's add a `custom_link` field in the tests too

Comment (DrkSephy) [reply datadog/tests/resource_datadog_dashboard_geomap_test.go:8]: Done in commit [cc356af](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/cc356af8aa7b6b51bb3dbb0c4ec4bbd2b707f181).

File: datadog/resource_datadog_dashboard.go [hunk 2]
```diff
@@ -4344,6 +4361,169 @@ func buildTerraformServiceLevelObjectiveDefinition(datadogDefinition datadogV1.S
 	return terraformDefinition
 }
 
+//
+// Geomap Widget Definition helpers
+//
+func getGeomapDefinitionSchema() map[string]*schema.Schema {
```
[the remark this answers is not in the document]
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go:4367]: Done in commit [cc356af](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/cc356af8aa7b6b51bb3dbb0c4ec4bbd2b707f181).

File: datadog/resource_datadog_dashboard.go [hunk 3]
```diff
@@ -4344,6 +4361,169 @@ func buildTerraformServiceLevelObjectiveDefinition(datadogDefinition datadogV1.S
...
+						Type:        schema.TypeBool,
+						Required:    true,
+					},
+				},
+			},
+		},
+		"view": {
+			Description: "The view of the world that the map should render.",
+			Type:        schema.TypeList,
+			Required:    true,
```
[the remark this answers is not in the document]
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go:4400]: Done in commit [cc356af](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/cc356af8aa7b6b51bb3dbb0c4ec4bbd2b707f181).

File: datadog/resource_datadog_dashboard.go [hunk 4]
```diff
@@ -4344,6 +4361,169 @@ func buildTerraformServiceLevelObjectiveDefinition(datadogDefinition datadogV1.S
...
+			Type:        schema.TypeList,
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: getGeomapRequestSchema(),
+			},
+		},
+		"style": {
+			Description: "Style of the widget graph. One nested block is allowed with the structure below.",
+			Type:        schema.TypeList,
+			Optional:    true,
```
[the remark this answers is not in the document]
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go:4380]: Done in commit [cc356af](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/cc356af8aa7b6b51bb3dbb0c4ec4bbd2b707f181).

**Review by jbenais — APPROVED** [review]
LGTM 🚀

File: datadog/resource_datadog_dashboard.go [hunk 5]
```diff
@@ -4344,6 +4361,183 @@ func buildTerraformServiceLevelObjectiveDefinition(datadogDefinition datadogV1.S
...
+			Type:        schema.TypeString,
+			Optional:    true,
+		},
+		"title_align": {
+			Description:  "The alignment of the widget's title. One of `left`, `center`, or `right`.",
+			Type:         schema.TypeString,
+			ValidateFunc: validators.ValidateEnumValue(datadogV1.NewWidgetTextAlignFromValue),
+			Optional:     true,
+		},
+		"time":      getDeprecatedTimeSchema(),
```
Comment (nmuesch) [inline datadog/resource_datadog_dashboard.go:4428]: I think we should remove this line. As I understand, `live_span` covers the same functionality, but without the TypeMap deprecated path.

File: datadog/resource_datadog_dashboard.go [hunk 6]
```diff
@@ -4344,6 +4361,183 @@ func buildTerraformServiceLevelObjectiveDefinition(datadogDefinition datadogV1.S
...
+		datadogDefinition.SetTitleAlign(datadogV1.WidgetTextAlign(v))
+	}
+
+	if v, ok := terraformDefinition["time"].(map[string]interface{}); ok && len(v) > 0 {
+		datadogDefinition.Time = buildDatadogWidgetTime(v)
+	} else if ls, ok := terraformDefinition["live_span"].(string); ok && ls != "" {
+		datadogDefinition.Time = &datadogV1.WidgetTime{
+			LiveSpan: datadogV1.WidgetLiveSpan(ls).Ptr(),
+		}
+	}
```
Comment (nmuesch) [inline datadog/resource_datadog_dashboard.go:4491]: If we decide to remove the `time` part of the schema, we also need to remove this and just keep whats in the `else`

File: datadog/resource_datadog_dashboard.go [hunk 7]
```diff
@@ -5060,6 +5254,49 @@ func buildDatadogTraceServiceDefinition(terraformDefinition map[string]interface
...
+		terraformDefinition["title_align"] = *v
+	}
+
+	if v, ok := datadogDefinition.GetTimeOk(); ok {
+		// Set to deprecated field if that's what is used in the config, otherwise, set in the new field
+		if _, ok := k.GetOkWith("time"); ok {
+			terraformDefinition["time"] = buildTerraformWidgetTimeDeprecated(*v)
+		} else {
+			terraformDefinition["live_span"] = v.GetLiveSpan()
+		}
```
Comment (nmuesch) [inline datadog/resource_datadog_dashboard.go:5295]: Same here

**Review by nmuesch — COMMENTED** [review]
Overall looks good, I left some inline comments, I don't think we should introduce a deprecated field on new widgets. 

Otherwise, can you pull in the latest master branch? There are new CI additions for docs and linting that would be good to let run before we merge.

File: datadog/resource_datadog_dashboard.go (hunk 7 above)
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go:5291]: Done in commit [82940d4](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/82940d4ce6e86936c65dd22e85b98ba68bcc21ef).

File: datadog/resource_datadog_dashboard.go (hunk 6 above)
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go]: Done in commit [82940d4](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/82940d4ce6e86936c65dd22e85b98ba68bcc21ef).

File: datadog/resource_datadog_dashboard.go (hunk 5 above)
Comment (DrkSephy) [reply datadog/resource_datadog_dashboard.go]: Done in commit [82940d4](https://github.com/DataDog/terraform-provider-datadog/pull/954/commits/82940d4ce6e86936c65dd22e85b98ba68bcc21ef).

**Review by nmuesch — APPROVED** [review]

**Pull request merged**